File: backend/app/api/routers/assets.py
from __future__ import annotations

from fastapi import APIRouter, Depends, Query

from app.api.deps import Principal, get_assets_repo, get_current_principal
from app.repositories.assets import AssetsRepository
from app.schemas.profile import AssetSummary

router = APIRouter(prefix="/api/assets", tags=["assets"])


@router.get("/search", response_model=list[AssetSummary])
async def search_assets(
    q: str = Query(min_length=1, max_length=50, description="Ticker prefix or part of the name"),
    limit: int = Query(8, ge=1, le=20),
    _: Principal = Depends(get_current_principal),
    assets: AssetsRepository = Depends(get_assets_repo),
) -> list[dict]:
    """Type-ahead for picking tickers to follow."""
    return await assets.search(q, limit)

File: frontend/src/services/assets.ts
import { api } from "@/lib/api";
import type { AssetSummary } from "@/lib/types";

/** The recommendable stocks and ETFs (backend: app/api/routers/assets.py). */
export const assetsService = {
  /** Type-ahead by ticker prefix or name. Silent: no full-screen overlay while typing. */
  search: (query: string) => api<AssetSummary[]>(`/api/assets/search?q=${encodeURIComponent(query)}&limit=8`, { silent: true }),
};
